squire_lib/src/ffi_tournament.rs

```diff
@@ -86,6 +123,173 @@ impl TournamentId {
+        }
+
+        if starting_table_number != tournament.round_reg.starting_table {
+            op_vect.push(TournOp::UpdateTournSetting(
+                aid,
+                TournamentSetting::StartingTableNumber(starting_table_number),
+            ));
+        }
+
+        if use_table_number != tournament.use_table_number {
+            op_vect.push(TournOp::UpdateTournSetting(
+                aid,
+                TournamentSetting::UseTableNumbers(use_table_number),
+            ));
+        }
+
+        if game_size != tournament.game_size {
+            op_vect.push(TournOp::UpdateTournSetting(
+                aid,
+                TournamentSetting::PairingSetting::MatchSize(game_size),
+            ));
+        }
+
+        if min_deck_count != tournament.min_deck_count {
+            op_vect.push(TournOp::UpdateTournSetting(
+                aid,
+                TournamentSetting::MinDeckCount(min_deck_count),
+            ));
+        }
+
+        if max_deck_count != tournament.max_deck_count {
+            op_vect.push(TournOp::UpdateTournSetting(
+                aid,
+                TournamentSetting::MaxDeckCount(max_deck_count),
+            ));
+        }
+
+        if reg_open != tournament.reg_open {
+            op_vect.push(TournOp::UpdateReg(aid, reg_open));
+        }
+
+        if require_check_in != tournament.require_check_in {
+            op_vect.push(TournOp::UpdateTournSetting(
+                aid,
+                TournamentSetting::RequireCheckIn(require_check_in),
+            ));
+        }
+
+        if require_deck_reg != tournament.require_deck_reg {
+            op_vect.push(TournOp::UpdateTournSetting(
+                aid,
+                TournamentSetting::RequireDeckReg(require_deck_reg),
+            ));
+        }
+
+        // Apply all settings
+        let err: bool = false;
+        for i in 0..op_vect.len() {
+            match tournament.apply_op(op_vect[i]) {
+                Err(t_err) => {
+                    println!("[FFI]: update_settings error: {t_err}");
+                    err = true;
+                    break;
+                }
+                _ => {}
+            }
+        }
+
+        // Rollback on error.
+        if err {
+            // Panic on double trouble
+        }
+
+        return true;
+    }
+
+    /// Pairs a set of rounds
+    /// Returns a null terminated list of the round ids
+    /// Returns NULL on error
+    #[no_mangle]
+    pub unsafe extern "C" fn tid_pair_round(self: Self) -> *const RoundId {
+        let admin_account: AdminId = Uuid::default().into();
+        let op: TournOp = TournOp::PairRound(admin_account);
+        match FFI_TOURNAMENT_REGISTRY.get().unwrap().get_mut(&self) {
+            Some(mut t) => {
+                match t.apply_op(op) {
+                    Ok(Pair(ident_vec)) => {
+                        let len: usize = (ident_vec.len() + 1) * std::mem::size_of::<RoundId>();
+                        let ptr = System
```

Comment:
Can we generalize this? Having repeat code like this is less than ideal. Having repeat code that does pointer operations seems like a bad idea. Something like this:
```rust
pub fn copy_to_system_pointer<T: Default, I: ExactLengthIterator<Item = T>>(iter: I) -> *const T {
    let len = (iter.len() + 1) * std::mem::size_of::<T>();
    let ptr = System
            .allocate(Layout::from_size_align(len, 1).unwrap())
            .unwrap()
            .as_mut_ptr() as *mut T;
    let slice = unsafe { &mut *(ptr::slice_from_raw_parts(ptr, len) as *mut [T]) };
    slice.iter_mut().zip(iter).for_each(|(dst, p)| { *dst = *p; });
    slice[len-1] = T::default().into();
    ptr
}
```

Reply:
Perhaps this should be marked as an `unsafe` function as it requires the caller to pass a type `T` whose default is null.

Reply:
The last slice should always be nulled for the entire size, relying on the default of T seems wrong, the best way to do this in C++ would be a `memset(ptr + len - 1, 0, sizeof(T))` or something; I am not sure if I can do something similar in rust.

Reply:
Most primitive types (or near-primitive, like `Uuid`) have defaults that are null. This method is meant to pass data across the FFI boundary, so I doubt we will be using much more than `Uuid`s and primitives.

Marking a function as `unsafe` signals to the caller that there is some invariant that they are in charge of upholding (that is literally what an `unsafe` function means). We should label that as such and document that fact.

Alternatively, we can try to allocate nulled memory or use `size_of` again to null out the last few bytes.

At any rate, a function like this should exist.